### PR TITLE
[Merged by Bors] - feat(Algebra): integer powers of -1

### PR DESCRIPTION
**Pull request by joelriou**
This PR introduces the map `Int.negOnePow : ℤ → ℤ` which sends `n` to `(-1 : Units ℤ) ^ n`.

---
@jcommelin: I hope it is ok for you to appear in the `Authors:` list of this file: the name `neg_one_pow` and some lemmas were in the code you contributed to the LTE.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by linesthatinterlace** [comment]
Is this map not in fact a group hom (albeit one from an additive group to a multiplicative group)? Specifically, it is the homomorphism from Z to Units Z whose kernel is the even integers. I suppose this is Z to Additive (Units Z), maybe, as an AddMonoidHom?

I think we should have this fact (and ideally in fact define it as such an object).

**Comment by alreadydone** [comment]
> Is this map not in fact a group hom (albeit one from an additive group to a multiplicative group)? Specifically, it is the homomorphism from Z to Units Z whose kernel is the even integers. I suppose this is Z to Additive (Units Z), maybe, as an AddMonoidHom?

> I think we should have this fact (and ideally in fact define it as such an object).

There is [zpowersHom](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/GroupPower/Lemmas.html#zpowersHom); maybe we can develop more general API around it connecting to [orderOf](https://leanprover-community.github.io/mathlib4_docs/Mathlib/GroupTheory/OrderOfElement.html#orderOf).

**Comment by joelriou** [comment]
In order to clarify the intent of this PR, its purpose is to have a convenient way to manipulate `(-1)^n` in computations (e.g. signs in (co)homology theory). Obviously, there is an intermediate isomorphism between the integers mod two and the multiplicative group of the integers which could be used here, but it would add two layers (the surjective map from the integers to the integers mod 2, and the inclusion of units). I have attempted to use the unfolded syntax `((-1 : Units ℤ) ^ n : ℤ)`, but doing computations with it was not very practical. Whichever way it is done internally, I think it is good to have a definition `Int.newOnePow : ℤ → ℤ` like it was done in the Liquid Tensor Experiment.

File: Mathlib/Algebra/GroupPower/NegOnePow.lean [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+This file defines the map `negOnePow : ℤ → ℤ` which sends `n` to `(-1 : Units ℤ) ^ n`.
+
+The definition of `negOnePow` and some lemmas first appeared in contributions by
+Johan Commelin to the Liquid Tensor Experiment.
+
+-/
+
+namespace Int
+
+/-- The map `ℤ → ℤ` which sends `n` to `(-1 : Units ℤ) ^ n`. --/
```
Comment (jcommelin) [inline Mathlib/Algebra/GroupPower/NegOnePow.lean:20]: Two minor edits:
```suggestion
/-- The map `ℤ → ℤ` which sends `n` to `(-1 : ℤˣ) ^ n`. -/
```

File: Mathlib/Algebra/GroupPower/NegOnePow.lean [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    simp only at h
+  · exact negOnePow_odd n
+
+@[simp]
+lemma negOnePow_neg (n : ℤ) : (-n).negOnePow = n.negOnePow := by
+  dsimp [negOnePow]
+  simp only [zpow_neg, ← inv_zpow, inv_neg, inv_one]
+
+lemma negOnePow_sub (n₁ n₂ : ℤ) :
+    (n₁ - n₂).negOnePow =  n₁.negOnePow * n₂.negOnePow := by
```
Comment (jcommelin) [inline Mathlib/Algebra/GroupPower/NegOnePow.lean:76]:
```suggestion
    (n₁ - n₂).negOnePow = n₁.negOnePow * n₂.negOnePow := by
```

File: Mathlib/Algebra/GroupPower/NegOnePow.lean [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+Copyright (c) 2023 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou, Johan Commelin
+-/
+import Mathlib.Data.Int.Parity
+
+/-!
+# Integer powers of (-1)
+
+This file defines the map `negOnePow : ℤ → ℤ` which sends `n` to `(-1 : Units ℤ) ^ n`.
```
Comment (jcommelin) [inline Mathlib/Algebra/GroupPower/NegOnePow.lean:11]:
```suggestion
This file defines the map `negOnePow : ℤ → ℤ` which sends `n` to `(-1 : ℤˣ) ^ n`.
```

File: Mathlib/Algebra/GroupPower/NegOnePow.lean [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+The definition of `negOnePow` and some lemmas first appeared in contributions by
+Johan Commelin to the Liquid Tensor Experiment.
+
+-/
+
+namespace Int
+
+/-- The map `ℤ → ℤ` which sends `n` to `(-1 : Units ℤ) ^ n`. --/
+@[pp_dot]
+def negOnePow (n : ℤ) : ℤ := (-1 : Units ℤ) ^ n
```
Comment (jcommelin) [inline Mathlib/Algebra/GroupPower/NegOnePow.lean:22]:
```suggestion
def negOnePow (n : ℤ) : ℤ := (-1 : ℤˣ) ^ n
```

File: Mathlib/Algebra/GroupPower/NegOnePow.lean [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+lemma negOnePow_def (n : ℤ) : n.negOnePow = (-1 : Units ℤ) ^ n := rfl
```
Comment (jcommelin) [inline Mathlib/Algebra/GroupPower/NegOnePow.lean:24]:
```suggestion
lemma negOnePow_def (n : ℤ) : n.negOnePow = (-1 : ℤˣ) ^ n := rfl
```

**Review by jcommelin — APPROVED** [review]
The stuff in LTE convinced me that this is a useful approach. We might figure out a better approach at some point, but for now, let's have this.

bors d+

**Comment by joelriou** [comment]
Thanks very much for the reviews!

bors merge

**Comment by eric-wieser** [comment]
> I have attempted to use the unfolded syntax `((-1 : Units ℤ) ^ n : ℤ)`, but doing computations with it was not very practical.

Can you elaborate on this? I effectively rejected a similar PR for mathlib 3, https://github.com/leanprover-community/mathlib/pull/11204, because I argued we should instead be filling out the API to make it more practical (which I believe we largely did)

**Pull request closed without merging**